[SPARK-29585][WEBUI] Fix stagePage duration

teeyog opened the pull request:
### What changes were proposed in this pull request?
This pr proposes  to fix the Duration  in Completed Stages in StagePage in the spark UI.

JIRA Issue: https://issues.apache.org/jira/browse/SPARK-29585

### Why are the changes needed?
 Summary Metrics for Completed Tasks uses ```executorRunTime```, and Duration in Task details uses ```executorRunTime```, but Duration in Completed Stages uses ```stageData.completionTime - stageData.firstTaskLaunchedTime```, which results in different results. So the Duration in Completed Stages in StagePage should also use executorRunTime.

Commit:
jdbc dialect can match non-lowercase URL prefixes

Commit:
jdbc dialect matching is not case sensitive

Commit:
Merge branch 'master' of https://github.com/apache/spark into sql_dialect

Commit:
jdbc dialect matching is not case sensitive

Commit:
jdbc dialect matching is not case sensitive

Commit:
jdbc dialect matching is not case sensitive

Commit:
jdbc dialect matching is not case sensitive

Commit:
jdbc dialect matching is not case sensitive

Commit:
jdbc dialect matching is not case sensitive

Commit:
Merge branch 'master' of https://github.com/apache/spark into sql_dialect

Commit:
Merge branch 'master' of https://github.com/apache/spark into sql_dialect

Commit:
fix stagePage duration

Commit:
fix stagePage duration

Commit:
fix stagePage duration

AmplabJenkins commented:
Can one of the admins verify this patch?

uncleGen commented:
IIUC, the `executorRunTime` means the time spent to run one task. The stage duration is the from time first task launched to time last task completed, i.e. the stage completion time. These two duration is different.

teeyog commented:
@uncleGen Thank you for your reply, I would like to ask the stage duration is the from time first task launched to time last task completed, why the task duration does not use the completed time - launched time, I think the calculation of duration in all places should be the same.

srowen commented:
@teeyog I dont' quite understand, isn't that what this does already? see the comments you deleted too

teeyog commented:
@srowen @uncleGen The phenomenon is: If a stage has only one task, then the duration displayed in `Metrics for 1 Completed Tasks` should be the same as the duration of the specific task below, and same as duration in stagePage, but it is different now, because the first two are using executorRunTime , while stagePage uses ```completed time - launched time```

teeyog commented:
![image](https://user-images.githubusercontent.com/19678557/67611174-674c0300-f7ca-11e9-9d24-a49f56ae2cf3.png)

![image](https://user-images.githubusercontent.com/19678557/67611782-433ef080-f7cf-11e9-8dfe-cbc3ce7663dc.png)
@srowen @uncleGen   These three data should be the same.

Commit:
jdbc dialect matching is not case sensitive

Commit:
jdbc dialect matching is not case sensitive

shahidki31 commented:
https://github.com/apache/spark/pull/26241#issuecomment-546554316

@teeyog Which version did you test? In master it seems not reproducible. Could you test in master?
![Screenshot 2019-10-26 at 3 45 12 PM](https://user-images.githubusercontent.com/23054875/67618284-e8c38580-f80a-11e9-9732-fecb052f5718.png)

Commit:
jdbc dialect matching is not case sensitive

teeyog commented:
@shahidki31

teeyog commented:
@shahidki31  This is the latest code of the master
![image](https://user-images.githubusercontent.com/19678557/67663849-59040f80-f9a1-11e9-98ee-5487a6aea785.png)

![image](https://user-images.githubusercontent.com/19678557/67664004-b39d6b80-f9a1-11e9-9fa7-be503454a34f.png)

shahidki31 commented:
Thanks @teeyog . This was the behavior earlier also right? ( I meant earlier versions of spark).

srowen commented on core/src/main/scala/org/apache/spark/ui/jobs/StageTable.scala:
```diff
@@ -417,19 +417,8 @@ private[ui] class StageDataSource(
-      }
-    }
+
+    val duration = Some(stageData.executorRunTime)
```
Just to be clear, this is exactly the field used to compute min/max stats, etc?

shahidki31 replied:
I think no. `stageData.executorRunTime` represents summation of `executorRunTime` of every task corresponding to the stage. Tasks need not be execute in sequential order (can be parallel also). So, I think this might not give the correct duration.

I think, field used to compute min/max stats, etc are `taskMetrics.executorRunTime`. @teeyog please correct me if I am wrong.

teeyog replied:
> I think no. `stageData.executorRunTime` represents summation of `executorRunTime` of every task corresponding to the stage. Tasks need not be execute in sequential order (can be parallel also). So, I think this might not give the correct duration.
> 
> I think, field used to compute min/max stats, etc are `taskMetrics.executorRunTime`. @teeyog please correct me if I am wrong.

What you said is correct. When the task is executed in parallel, the result is wrong. Then I think the task duration should not use executorRunTime. As shown in the previous figure, their values ​​are not match.

teeyog replied:
Do you think I am right, if I am going to develop

shahidki31 replied:
Thanks @teeyog. I think this was the behavior in the previous versions of spark. I am not sure we need to change it. What ever you are proposing is `duration = executorRunTime + ser/deser time + gettingResultTime`. But, Already these 3 columns are getting disaplyed in the task table right? If we unify all would we need to display individual columns?

teeyog replied:
@shahidki31 I think the task duration should be calculated using the ```completed time - launched time``` instead of the executorRunTime. This ensures that a stage has only one task, and the stage duration and task duration are the same.

shahidki31 replied:
Yes. `completed time - launched time = executorRunTime + ser/deser time + gettingResultTime`. right? Also, I think these 3 columns are getting displayed in the task table, Right?. If you really want, you may can add a new column and you can name it  `totalTime` which display the summation of 3. But still, I am not sure that will match with stage duration, as stage may can have other parameters for duration.

teeyog replied:
@shahidki31 
Yes, I think this can match with stage duration

jiangxb1987 committed:
Prepare Spark release v3.0.0-preview-rc2

### What changes were proposed in this pull request?

To push the built jars to maven release repository, we need to remove the 'SNAPSHOT' tag from the version name.

Made the following changes in this PR:
* Update all the `3.0.0-SNAPSHOT` version name to `3.0.0-preview`
* Update the sparkR version number check logic to allow jvm version like `3.0.0-preview`

**Please note those changes were generated by the release script in the past, but this time since we manually add tags on master branch, we need to manually apply those changes too.**

We shall revert the changes after 3.0.0-preview release passed.

### Why are the changes needed?

To make the maven release repository to accept the built jars.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

N/A

Commit:
jdbc dialect matching is not case sensitive

Commit:
jdbc dialect matching is not case sensitive

Commit:
dialect

Commit:
Merge branch 'sql_dialect' of https://github.com/teeyog/spark into sql_dialect

Commit:
jdbc dialect matching is not case sensitive

Commit:
fix

yaooqinn committed:
[SPARK-29653][SQL] Fix MICROS_PER_MONTH in IntervalUtils

### What changes were proposed in this pull request?

MICROS_PER_MONTH = DAYS_PER_MONTH * MICROS_PER_DAY

### Why are the changes needed?

fix bug

### Does this PR introduce any user-facing change?

no
### How was this patch tested?

add ut

Closes #26321 from yaooqinn/SPARK-29653.

Authored-by: Kent Yao <yaooqinn@hotmail.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

maropu committed:
[SPARK-29120][SQL][TESTS] Port create_view.sql

### What changes were proposed in this pull request?

This PR ports create_view.sql from PostgreSQL regression tests https://github.com/postgres/postgres/blob/REL_12_STABLE/src/test/regress/sql/create_view.sql

The expected results can be found in the link: https://github.com/postgres/postgres/blob/REL_12_STABLE/src/test/regress/expected/create_view.out

### Why are the changes needed?

To check behaviour differences between Spark and PostgreSQL

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

Pass the Jenkins. And, Comparison with PgSQL results

Closes #26290 from maropu/SPARK-29120.

Authored-by: Takeshi Yamamuro <yamamuro@apache.org>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

dongjoon-hyun committed:
[SPARK-29668][DOCS] Deprecate Python 3 prior to version 3.6

### What changes were proposed in this pull request?

This PR aims to deprecate `Python 3.4 ~ 3.5`, which is prior to version 3.6 additionally.

### Why are the changes needed?

Since `Python 3.8` is already out, we will focus on to support Python 3.6/3.7/3.8.

### Does this PR introduce any user-facing change?

Yes. It's highly recommended to use Python 3.6/3.7. We will verify Python 3.8 before Apache Spark 3.0.0 release.

### How was this patch tested?

NA (This is a doc-only change).

Closes #26326 from dongjoon-hyun/SPARK-29668.

Authored-by: Dongjoon Hyun <dhyun@apple.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

jiangxb1987 committed:
[SPARK-29646][BUILD] Allow pyspark version name format `${versionNumber}-preview` in release script

### What changes were proposed in this pull request?

Update `release-build.sh`, to allow pyspark version name format `${versionNumber}-preview`, otherwise the release script won't generate pyspark release tarballs.

### How was this patch tested?

Tested locally.

Closes #26306 from jiangxb1987/buildPython.

Authored-by: Xingbo Jiang <xingbo.jiang@databricks.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

jiangxb1987 committed:
[SPARK-29666][BUILD] Fix the publish release failure under dry-run mode

### What changes were proposed in this pull request?

`release-build.sh` fail to publish release under dry run mode with the following error message:
```
/opt/spark-rm/release-build.sh: line 429: pushd: spark-repo-g4MBm/org/apache/spark: No such file or directory
```

We need to at least run the `mvn clean install` command once to create the `$tmp_repo` path, but now those steps are all skipped under dry-run mode. This PR fixes the issue.

### How was this patch tested?

Tested locally.

Closes #26329 from jiangxb1987/dryrun.

Authored-by: Xingbo Jiang <xingbo.jiang@databricks.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

jiangxb1987 committed:
Revert "Prepare Spark release v3.0.0-preview-rc2"

This reverts commit 007c873ae34f58651481ccba30e8e2ba38a692c4.

rdblue committed:
[SPARK-29277][SQL] Add early DSv2 filter and projection pushdown

### What changes were proposed in this pull request?

This adds a new rule, `V2ScanRelationPushDown`, to push filters and projections in to a new `DataSourceV2ScanRelation` in the optimizer. That scan is then used when converting to a physical scan node. The new relation correctly reports stats based on the scan.

To run scan pushdown before rules where stats are used, this adds a new optimizer override, `earlyScanPushDownRules` and a batch for early pushdown in the optimizer, before cost-based join reordering. The other early pushdown rule, `PruneFileSourcePartitions`, is moved into the early pushdown rule set.

This also moves pushdown helper methods from `DataSourceV2Strategy` into a util class.

### Why are the changes needed?

This is needed for DSv2 sources to supply stats for cost-based rules in the optimizer.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

This updates the implementation of stats from `DataSourceV2Relation` so tests will fail if stats are accessed before early pushdown for v2 relations.

Closes #25955 from rdblue/move-v2-pushdown.

Authored-by: Ryan Blue <blue@apache.org>
Signed-off-by: Ryan Blue <blue@apache.org>

MaxGekk committed:
[SPARK-29651][SQL] Fix parsing of interval seconds fraction

### What changes were proposed in this pull request?
In the PR, I propose to extract parsing of the seconds interval units to the private method `parseNanos` in `IntervalUtils` and modify the code to correctly parse the fractional part of the seconds unit of intervals in the cases:
- When the fractional part has less than 9 digits
- The seconds unit is negative

### Why are the changes needed?
The changes are needed to fix the issues:
```sql
spark-sql> select interval '10.123456 seconds';
interval 10 seconds 123 microseconds
```
The correct result must be `interval 10 seconds 123 milliseconds 456 microseconds`
```sql
spark-sql> select interval '-10.123456789 seconds';
interval -9 seconds -876 milliseconds -544 microseconds
```
but the whole interval should be negated, and the result must be `interval -10 seconds -123 milliseconds -456 microseconds`, taking into account the truncation to microseconds.

### Does this PR introduce any user-facing change?
Yes. After changes:
```sql
spark-sql> select interval '10.123456 seconds';
interval 10 seconds 123 milliseconds 456 microseconds
spark-sql> select interval '-10.123456789 seconds';
interval -10 seconds -123 milliseconds -456 microseconds
```

### How was this patch tested?
By existing and new tests in `ExpressionParserSuite`.

Closes #26313 from MaxGekk/fix-interval-nanos-parsing.

Authored-by: Maxim Gekk <max.gekk@gmail.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

Commit:
Merge remote-tracking branch 'upstream/master' into sql_dialect

d80tb7 committed:
[SPARK-29126][PYSPARK][DOC] Pandas Cogroup udf usage guide

This PR adds some extra documentation for the new Cogrouped map Pandas udfs.  Specifically:

- Updated the usage guide for the new `COGROUPED_MAP` Pandas udfs added in https://github.com/apache/spark/pull/24981
- Updated the docstring for pandas_udf to include the COGROUPED_MAP type as suggested by HyukjinKwon in https://github.com/apache/spark/pull/25939

Closes #26110 from d80tb7/SPARK-29126-cogroup-udf-usage-guide.

Authored-by: Chris Martin <chris@cmartinit.co.uk>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
match stage duration and task duration

Commit:
match stage duration and task duration

Commit:
Merge branch 'master' of https://github.com/apache/spark

Commit:
match stage duration and task duration

Commit:
match stage duration and task duration

Commit:
match stage duration and task duration

Commit:
match stage duration and task duration

Commit:
Merge branch 'master' of https://github.com/apache/spark

Commit:
Merge branch 'spark_ui_duration' of https://github.com/teeyog/spark into spark_ui_duration

Commit:
match stage duration and task duration

Commit:
Merge branch 'spark_ui_duration' of https://github.com/teeyog/spark into spark_ui_duration

Commit:
match stage duration and task duration
